### ISSUE TITLE
self referencing data providers

### ISSUE THREAD

**Issue by theofidry** [sebastianbergmann/phpunit#3493]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 7.5.2
| PHP version         | 7.2+
| Installation Method | Composer & PHAR

Given the following test:

<details>
<summary>DemoTest.php</summary>

```php
<?php

declare(strict_types=1);

namespace Acme;

use function file_put_contents;
use Generator;
use const PHP_EOL;
use PHPUnit\Framework\TestCase;
use function spl_object_id;

class DemoTest extends TestCase
{

    /**
     * {@inheritdoc}
     */
    protected function setUp(): void
    {
        file_put_contents('php://stderr', PHP_EOL.'SETUP'.PHP_EOL);
        file_put_contents('php://stderr', spl_object_id($this).PHP_EOL);
    }

    /**
     * @dataProvider provideCallable
     */
    public function test_something(callable $callable): void
    {
        file_put_contents('php://stderr', PHP_EOL.'TEST'.PHP_EOL);
        file_put_contents('php://stderr', spl_object_id($this).PHP_EOL);

        $callable();

        $this->assertTrue(true);
    }

    public function provideCallable(): Generator
    {
        file_put_contents('php://stderr', PHP_EOL.'PROVIDER'.PHP_EOL);
        file_put_contents('php://stderr', spl_object_id($this).PHP_EOL);

        yield [
            function () {
                file_put_contents('php://stderr', PHP_EOL.'YIELDED CALLABLE'.PHP_EOL);
                file_put_contents('php://stderr', spl_object_id($this).PHP_EOL);
            }
        ];
    }
}
```
</details>

Will yield:

```
PROVIDER
336
SETUP
337
TEST
337
YIELDED CALLABLE
336
```

So you can see that the `$this` value reference changes between when the provider is called (hence when the `$this` is used in a callable as well) and when the test is set-up & executed.

It is worth mentioning that if providers are loaded lazily (not sure if it's gonna be done?) then the bug would be solved by itself. Otherwise maybe it could be worth recommending to write providers as static functions to avoid to reference`$this`.

**Comment by epdenouden** [comment]
> It is worth mentioning that if providers are loaded lazily (not sure if it's gonna be done?) then the bug would be solved by itself.

Hi @theofidry ! I am working on an implementation for lazy loaded data providers. It _can_ work, I am sure of that now. However, this change will have a lot of impact. Some test collections will certainly need a change to use it.

![image](https://user-images.githubusercontent.com/26651359/54877775-927d0700-4e23-11e9-88f8-3908930e9611.png)

The remaining 10% of the work is going to be 90% of the time, like it always is. But lazy Sundaymorning style dataproviders are coming for sure.

**Comment by theofidry** [comment]
Cool! Lemme know if you need me to try the branch or something :)

**Comment by epdenouden** [comment]
> Cool! Lemme know if you need me to try the branch or something :)

Thanks! Will do once the self-tests are green, so we have something to break horribly ;-)

```
./phpunit --debug tests/unit/Runner/TestSuiteSorterTest.php                                                                           ✔  14:16:48 
adding DataProviderTestSuite(PHPUnit\Runner\TestSuiteSorterTest, testShouldNotAffectEmptyTestSuite)
adding DataProviderTestSuite(PHPUnit\Runner\TestSuiteSorterTest, testBasicExecutionOrderOptions)
adding DataProviderTestSuite(PHPUnit\Runner\TestSuiteSorterTest, testOrderDurationWithoutCache)
adding DataProviderTestSuite(PHPUnit\Runner\TestSuiteSorterTest, testOrderDurationWithCache)
adding DataProviderTestSuite(PHPUnit\Runner\TestSuiteSorterTest, testSuiteSorterDefectsOptions)
PHPUnit 8.1-gdd1b6cf53 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.14 with Xdebug 2.6.0
Configuration: /Users/ewout/proj/phpunit/phpunit.xml

Test 'PHPUnit\Runner\TestSuiteSorterTest::testThrowsExceptionWhenUsingInvalidOrderOption' started
Test 'PHPUnit\Runner\TestSuiteSorterTest::testThrowsExceptionWhenUsingInvalidOrderOption' ended
Test 'PHPUnit\Runner\TestSuiteSorterTest::testThrowsExceptionWhenUsingInvalidOrderDefectsOption' started
Test 'PHPUnit\Runner\TestSuiteSorterTest::testThrowsExceptionWhenUsingInvalidOrderDefectsOption' ended
### should init dataprovider PHPUnit\Runner\TestSuiteSorterTest::testShouldNotAffectEmptyTestSuite here before run() checks
Test 'PHPUnit\Runner\TestSuiteSorterTest::testShouldNotAffectEmptyTestSuite with data set #0 (0, false, 0)' started
Test 'PHPUnit\Runner\TestSuiteSorterTest::testShouldNotAffectEmptyTestSuite with data set #0 (0, false, 0)' ended
Test 'PHPUnit\Runner\TestSuiteSorterTest::testShouldNotAffectEmptyTestSuite with data set #1 (0, false, 3)' started
Test 'PHPUnit\Runner\TestSuiteSorterTest::testShouldNotAffectEmptyTestSuite with data set #1 (0, false, 3)' ended
Test 'PHPUnit\Runner\TestSuiteSorterTest::testShouldNotAffectEmptyTestSuite with data set #2 (0, true, 0)' started
[etc etc]
```

Some of the side effects that need looking at:
- the `TestCase`s that make up a `DataProviderTestSuite` are not created, since there is no data to base them on
- the `DataProviderTestSuite` does not provide test counts before being initialized
- dataprovidered tests need to be available _before_ filters are applied

**Comment by rpkamp** [comment]
IIRC the problem with lazy loading data providers is that you can't tell beforehand any more how many tests there will be (the `Y` in the default `..... (X/Y)` printer output).

Any solution for that, or will the `Y` be dropped completely?

**Comment by epdenouden** [comment]
@rpkamp You are correct. Initializing the dataproviders in flight does make the test count inaccurate, but it doesn't seem to create other side effects.

**Comment by rpkamp** [comment]
Is there a possible refactoring that would make it less of a headache?

**Comment by epdenouden** [comment]
> Is there a possible refactoring that would make it less of a headache?

Yes, rewrite and I do not mean that jokingly. @sebastianbergmann is also rather done with the whole `TestSuite` thing which is also making `DataProviderTestSuite` even harder than it already is.

Seriously though the 'headache' is mostly me missing domain knowledge _and_ that knowledge being hard to find and absorb. I can write down on a piece of paper how a lazy loaded dataprovider would work in PHPUnit, but getting the freaking thing to actually work in the field is something else entirely.

Part of this is the whole `Iterator`-based setup for filtering which tests to run. Unfortunately the execution planning (=`foreach`ing over the collection of tests) is so intertwined with reading the configuration and the way `TestCase` objects are created from the actual PHP and PHPT source code. See also: the fun I had with getting better code location hints and execution reordering.

For example. Yesterday I got some experimental code working that actually loads and executed dataproviders on demand. Still fails with `--list-tests`, because that gets its own iterator, which doesn't load+filter the dataprovidered tests properly. Same difference with the execution reordering algorithm, which requires the full tree to already be there.

**Comment by epdenouden** [comment]
@rpkamp basically this :)

![image](https://user-images.githubusercontent.com/26651359/55076519-ee4ec680-5095-11e9-8e7d-cab1c1709f02.png)

**Comment by rpkamp** [comment]
So basically everything is being done quite ad-hoc, rather than in clearly defined stages that can be easily composed to allow different workflows?

**Comment by epdenouden** [comment]
Currently I am doing some experimentation on the current codebase with some exploratory coding. See what is (already) possible and where the internal boundaries lie. This is as much a result of me learning about all the various use-cases people use PHPUnit for.

There is a method to the madness, though! :-) I hope to meet @sebastianbergmann this spring and perhaps a clear roadmap springs into existence. My own work is focus on finding and improving/fixing the various boundaries inside PHPUnit. In theory every unit test is a unit, right? Pushing v7+v8 to their limits gives us lots of information about the multitude of edge cases people use the thing for.

In short: clearly defined stages that can be reused, or expanded, or refactored or whatevered would be great! And of course, have at it with experiments, PRs, etc.

**Comment by theofidry** [comment]
@sebastianbergmann I cannot see any PR links, to due what degree was this implemented? IIRC from the description the issue was the `$this` reference which changes which could be fixed by either lazy data providers or not allowing them to be non-static. Or is this closed as "won't fix"?

**Comment by sebastianbergmann** [comment]
Won't fix, yes.

**Comment by theofidry** [comment]
You would be open to a recommendation in the docs about making the data providers static then?
